### PR TITLE
Exempt if sender & target has same weight with exemption-luckperms addon

### PR DESCRIPTION
**Pull request by BlueTree242**
In most cases, An admin should not be able to punish another admin.

### CODE REVIEW

**Comment by A248** [comment]
This might be a breaking change, so I think there should be a configuration option for it. It will have to default to the old behavior.

During testing, you can set the configuration value via the mocking infrastructure.

**Comment by BlueTree242** [comment]
done :)

File: bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java [hunk 1]
```diff
@@ -19,7 +19,13 @@
 
 package space.arim.libertybans.core.addon.exempt.luckperms;
 
+import space.arim.dazzleconf.annote.ConfComments;
+import space.arim.dazzleconf.annote.ConfDefault;
 import space.arim.libertybans.core.addon.AddonConfig;
 
 public interface ExemptionLuckPermsConfig extends AddonConfig {
+
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java:27]:
```suggestion
    @ConfKey("exempt-same")
```

File: bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java [hunk 2]
```diff
@@ -19,7 +19,13 @@
...
+    @ConfComments("Whether to exempt punishment if operator & target have the same weight.")
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java:28]:
```suggestion
    @ConfComments("Whether to exempt victims with the same weight as the operator.")
```

File: bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java [hunk 3]
```diff
@@ -19,7 +19,13 @@
...
+    @ConfDefault.DefaultBoolean(false)
+    boolean exempt_same();
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java:30]:
```suggestion
    boolean exemptSame();
```

File: bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/LuckPermsExemptProvider.java [hunk 4]
```diff
@@ -73,14 +73,15 @@ public CompletionStage<Boolean> isExempted(CmdSender sender, String category, Vi
 		return userManager.loadUser(senderUuid).thenCombine(userManager.loadUser(targetUuid), (senderUser, targetUser) -> {
 			int senderWeight = calculateUserMaxWeight(senderUser);
 			int targetWeight = calculateUserMaxWeight(targetUser);
-			return targetWeight > senderWeight;
+			if (senderWeight == -1 && targetWeight == -1) return false;
+			return addon.config().exempt_same() ? targetWeight >= senderWeight : targetWeight > senderWeight;
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/LuckPermsExemptProvider.java:77]:
```suggestion
			return addon.config().exemptSame() ? targetWeight >= senderWeight : targetWeight > senderWeight;
```

File: bans-core-addons/exemption-luckperms/src/test/java/space/arim/libertybans/core/addon/exempt/luckperms/LuckPermsExemptProviderTest.java [hunk 5]
```diff
@@ -124,4 +125,19 @@ public void targetHasExemptWeightLowerThanOperatorWeight() {
...
+		setGroups(senderUser, null, null, 4, null, null, 5, 20);
+		setGroups(targetUser, 3, null, 10, null, 20, 9);
+		assertIsExempted(false);
+	}
+
+	@Test
+	public void targetHasExemptWeightSameAsOperatorWeightOptionEnabled() {
+		setGroups(senderUser, null, null, 4, null, null, 5, 20);
+		setGroups(targetUser, 3, null, 10, null, 20, 9);
+		when(config.exempt_same()).thenReturn(true);
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/test/java/space/arim/libertybans/core/addon/exempt/luckperms/LuckPermsExemptProviderTest.java:139]:
```suggestion
		when(config.exemptSame()).thenReturn(true);
```

File: bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java [hunk 6]
```diff
@@ -19,7 +19,13 @@
 
 package space.arim.libertybans.core.addon.exempt.luckperms;
 
+import space.arim.dazzleconf.annote.ConfComments;
+import space.arim.dazzleconf.annote.ConfDefault;
```
Comment (A248) [inline bans-core-addons/exemption-luckperms/src/main/java/space/arim/libertybans/core/addon/exempt/luckperms/ExemptionLuckPermsConfig.java:23]:
```suggestion
import space.arim.dazzleconf.annote.ConfDefault;
import space.arim.dazzleconf.annote.ConfKey;
```

**Review by A248 — APPROVED** [review]
Looks good. The method and option names should be tweaked to match Java and YAML convention, a minor matter which I will fix for you.

Thanks for the contribution :)

**Pull request merged**